Add new `execute` method for getting a Regexp MatchData

andrewvc opened the pull request:
This is in direct support of https://github.com/logstash-plugins/logstash-filter-grok/pull/88

We need the ability to execute the match / capture in two phases while maintaining performance.

andrewvc commented:
Maybe I should use a `yield` instead of the block in https://github.com/jordansissel/ruby-grok/blob/master/lib/grok-pure.rb#L188 since `yield` is faster. Thoughts? https://docs.omniref.com/ruby/2.2.0/symbols/Proc/yield#annotation=4087638&line=711&hn=1

jordansissel commented on lib/grok-pure.rb:
```diff
@@ -172,7 +172,16 @@ def match(text)
 
+  # Returns the matched regexp object directly for performance at the
+  # cost of usability
+  def match_direct(text)
```
I don't like the `direct` name, but am having a hard time coming up with a better name. 

What about calling this `execute` and documenting that it returns a MatchData on success or `nil` on failure?

We can refactor later the rest of the library to use `execute(text)` instead of invoking `@regexp.match(text)` directly as they do.

jordansissel commented:
> Maybe I should use a yield instead of the block

There's a typo/bug in this method. The method definition should be `def capture(match, &block)`. Can you fix that here, since the intent of this PR feels like separating the match + capture phases.

andrewvc commented:
@jordansissel agreed on the typo but what do you think of my question of simplying getting rid of `&block` and replacing it with yield to bring it to performance parity with `match_and_capture`?

andrewvc commented:
@jordansissel one thing we should keep in mind is that if that is a typo/bug then this will break that method's signature with no backward compat. I'm personally fine with it, but just sayin'

jordansissel commented:
> replacing it with yield to bring it to performance parity with match_and_capture

If the following method definition exhibits different performance characteristics, this seems like a bug in ruby:

```
def foo(&block)
   yield "Hello"
end

def foo(&block)
  block.call "Hello"
end
```

My comment about `&block` was about a typo in the current method definition:

```
def capture(match, block)
                   ^ Missing &, it should be &block
```

jordansissel commented:
> then this will break that method's signature with no backward compat

We wrote this method years ago, but never used it. Its method definition is written incorrectly and requires users to do basically this: `grok.capture(m, proc { |k, v| ... })` instead of passing a block to the method like this: `grok.capture(m) { |k,v| ... }`. IMO, this typo is a bug that has been present since `def capture` was first written, and went unnoticed because we never call the method. The bug should be fixed. I don't want to make backwards-compatibility promises for *bugs* ;)

andrewvc commented:
@jordansissel I think I've addressed all your concerns with my latest push.

Benchmark Results:
```
Rehearsal ----------------------------------------------------------
10m Named Captures On  180.930000   0.720000 181.650000 (176.236212)
10m Named Captures Off 240.880000   0.940000 241.820000 (233.876933)
----------------------------------------------- total: 423.470000sec

                             user     system      total        real
10m Named Captures On  180.050000   0.670000 180.720000 (175.083869)
10m Named Captures Off 242.380000   0.890000 243.270000 (235.008755)
~/p/l/ruby-grok (match_direct) $ gem install jls-grok-0.11.3.gem
Successfully installed jls-grok-0.11.3
1 gem installed
~/p/l/ruby-grok (match_direct) $ ruby test/pure-ruby/benchmark_pattern.rb
Rehearsal ----------------------------------------------------------
10m Named Captures On  149.730000   0.670000 150.400000 (145.569641)
10m Named Captures Off 189.870000   0.680000 190.550000 (183.904052)
----------------------------------------------- total: 340.950000sec

                             user     system      total        real
10m Named Captures On  144.800000   0.500000 145.300000 (140.756561)
10m Named Captures Off 197.790000   1.070000 198.860000 (192.762491)
```

andrewvc commented:
Ah, I missed the new comment on `execute` will change that and get back to you here

andrewvc commented:
@jordansissel ok, this is ready for a new round of review now!

jordansissel commented:
LGTM, haven't run tests though, I assume they pass ;P

andrewvc commented:
Thanks @jordansissel ! Yes, the tests all pass.
